modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/toast/commands/OpenToast.es.js

```diff
@@ -136,15 +121,16 @@ function openToast({
 			displayType={type}
 			onClick={(event) => onClick({event, onClose: onCloseFn})}
 			onClose={onCloseFn}
-			title={
-				title && (
-					<Text allowHTML={titleType === TYPES.HTML} string={title} />
-				)
-			}
 			variant={variant}
 			{...toastProps}
 		>
-			<Text allowHTML={messageType === TYPES.HTML} string={message} />
+			<div
+				dangerouslySetInnerHTML={{
+					__html: `${
+						title ? `<strong class="lead">${title}</strong>` : ''
+					}${message}`,
+				}}
```

Comment:
Code change looks fine to me. One thing though. At the moment I believe most providers of `title` are going to include a trailing whitespace, but that's a bit painful/noisy for them. We could harmlessly include whitespace between `title` and `message` as a convenience here instead. If they already include trailing whitespace in `title`, the browser will collapse it anyway. ie.

```js
title ? `<strong class="lead">${title}</strong> ` : ''
//                                     note:   ^
```

The only thing that I would want to be sure of is this: *are we sure that `message` and `title` are never coming from untrusted user input?*

That's why when [I originally listed](https://github.com/liferay-frontend/liferay-portal/pull/173#issuecomment-668443561) "always assume HTML" as a possibility, I said:

> just assume HTML always (would require an analysis of existing use cases — may be risky)

As long as we're sure that all the messages at all the call sites are coming from us, then 👍

Reply:
> One thing though. At the moment I believe most providers of title are going to include a trailing whitespace [...]

So, `.lead` already includes `margin-right: 0.3125rem`. I initially had the extra spacey thing and it made the message look off.

> The only thing that I would want to be sure of is this: are we sure that message and title are never coming from untrusted user input?
>
> As long as we're sure that all the messages at all the call sites are coming from us, then 👍

I'm sure they might not, and not sure we can do this at this level. 

I know you usually need to escape the closer to the context that you can, which initially would indeed indicate here which is where we're rendering the html. We don't have the knowledge about whether `message` and `title` are safe though, so I think one idea that was floated around in the past was to mark these as `unsafe` so that consumers of the API would know to escape if needed.

Most messages are simple enough and I don't seem to contain user-generated input on a quick inspection. In general, I see a lot of wrapping, though, so it's hard to track what content is being passed down the chain...

Reply:
> So, `.lead` already includes `margin-right: 0.3125rem`. I initially had the extra spacey thing and it made the message look off.

So we still have the trailing spaces in the callsites then ([example here](https://github.com/liferay-frontend/liferay-portal/blob/f2a006a711a5ca8e4ee96861f07badc87f6b0a20/portal-web/docroot/html/common/themes/top_js.jspf#L125)), which means we have spacing-in-text _and_ spacing-in-CSS. I think we need the spacing in the text to stop the screenreader from saying something like "titlemessage" when the markup is `<span>title:</span>message`, which is why I suggested having it here once instead of at the callsites. Maybe it only looked weird because you added it here but didn't remove it from the callsites?

> not sure we can do this at this level.

Yeah, I think by definition, when you have an API that takes input and renders HTML, you're already past the point where you can check whether or not you trust the input — it's a string at that point.

In practical terms, I guess we just have to go with your quick inspection, and maybe plaster a warning in the doc comment (that nobody will read) that the message and title are rendered raw and must come from a trusted source. Another option that might help is to rename the parameters to `messageHTML` and `titleHTML`, so at least the caller knows that they are going to be dumped as-is into the DOM; it's not a foolproof safeguard, but it does at least increase the chance that they _might_ think about passing in untrusted input. 🤷

Reply:
> So we still have the trailing spaces in the callsites then (example here), which means we have spacing-in-text and spacing-in-CSS. I think we need the spacing in the text to stop the screenreader from saying something like "titlemessage" when the markup is <span>title:</span>message, which is why I suggested having it here once instead of at the callsites. Maybe it only looked weird because you added it here but didn't remove it from the callsites?

That's a fair point. I actually remember looking at something similar when we did [Updates clay:alert and clay:stripe tag implementation and usages](https://github.com/brianchandotcom/liferay-portal/pull/90058/files). In that case, was making sure title messages didn't need the `:`.

I checked right now both with `ChromeVox` and `VoiceOver` on mac. 
- They both seem to take the `:` as an almost full stop, so when present, it does read as "title{full_stop}message".
- `VoiceOver` does by default a better job at separating the words. I suspect it might be dependant on their html grouping/containers.
- None seem to react to spaces. Adding any amount of spacing around the colon doesn't seem to change the pause between words (not that I did a proper scientific measurement).

I checked all call sites (kinda) and only a few manually add the `:` and most of them use the usual simple title keys: `succes`, `error`, `warning`... Based on that, and the fact that we already [do this for `clay:alert` and `clay:stripe`](https://github.com/liferay-frontend/liferay-portal/blob/master/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/AlertTag.java#L241), it would seem reasonable to:
- Always add the colon inside `openToast` as we generate the toast content
- Change call sites that are currently passing a colon so they don't pass one (forgivable breaking change, maybe...)
- Add a default `type` to `title` transformation to reduce boilerplate. We [already do this for `clay:alert` and `clay:stripe`](https://github.com/liferay-frontend/liferay-portal/blob/master/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/AlertTag.java#L270-L281).

Does this seem reasonable to you, @wincent?

> In practical terms, I guess we just have to go with your quick inspection, and maybe plaster a warning in the doc comment (that nobody will read) that the message and title are rendered raw and must come from a trusted source. Another option that might help is to rename the parameters to messageHTML and titleHTML, so at least the caller knows that they are going to be dumped as-is into the DOM; it's not a foolproof safeguard, but it does at least increase the chance that they might think about passing in untrusted input. 🤷

This would be a bigger breaking change though... we could keep both, deprecate the non-suffixed one and replace our usages... not sure how much this is worth, though 🤔

Reply:
> Does this seem reasonable to you, @wincent?

Yes, although funnily enough, I actually mistyped the example above; instead of:

> saying something like "titlemessage" when the markup is `<span>title:</span>message`

I meant to write:

> saying something like "titlemessage" when the markup is `<span>title</span>message`

The colon snuck in there by mistake. I'm not at all surprised that a screenreader reads "foo:bar" as "foo _pause_ bar".

> not sure how much this is worth, though

I don't either. The probability of this biting us in the form of a nasty security incident in the future is somewhere between 0 and 1, but likely to be closer to the 0 end of the spectrum. And the probability that renaming the param makes the mistake less likely is also somewhere between 0 and 1, but likely closer to the 0 end too... In other words, I think it's ok to leave it as it is, but I at least wanted to float the idea.

Reply:
Do you think there's some value in adding both `:` and ` ` like `<strong class="lead">${title}:</strong> ${message}`?

Reply:
I think if you're going to mess with it (eg. adding colon) you may as well do all the touching at once (ie. adding the space).